crates/red_knot_python_semantic/resources/mdtest/assignment/augmented.md

```diff
@@ -45,3 +45,107 @@ x -= 1
 # TODO: should error, once operand type check is implemented
 reveal_type(x)  # revealed: int
 ```
+
+## Method union
+
+```py
+def bool_instance() -> bool:
+    return True
+
+flag = bool_instance()
+
+class Foo:
+    if bool_instance():
+        def __iadd__(self, other: int) -> str:
+            return "Hello, world!"
+    else:
+        def __iadd__(self, other: int) -> int:
+            return 42
+
+f = Foo()
+f += 12
+
+reveal_type(f)  # revealed: str | int
+```
+
+## Partially bound `__iadd__`
+
+```py
+def bool_instance() -> bool:
+    return True
+
+class Foo:
+    if bool_instance():
+        def __iadd__(self, other: str) -> int:
+            return 42
+
+f = Foo()
+f += "Hello, world!"
+
+reveal_type(f)  # revealed: int
+```
+
+## Partially-bound with `__add__`
+
+```py
+def bool_instance() -> bool:
+    return True
+
+class Foo:
+    def __add__(self, other: int) -> str:
+        return "Hello, world!"
+    if bool_instance():
+        def __iadd__(self, other: str) -> int:
+            return 42
+
+f = Foo()
+f += "Hello, world!"
+
+# TODO(charlie): This should be `int | str`, since `__iadd__` may be unbound.
+reveal_type(f)  # revealed: int
+```
+
+## Partially-bound target union
+
+```py
+def bool_instance() -> bool:
+    return True
+
+class Foo:
+    def __add__(self, other: int) -> str:
+        return "Hello, world!"
+    if bool_instance():
+        def __iadd__(self, other: str) -> int:
+            return 42
+
+if bool_instance():
+    f = Foo()
+else:
+    f = 42.0
+f += 12
+
+# TODO(charlie): This should be `str | int | float`
+reveal_type(f)  # revealed: @Todo
+```
+
+## Target union
+
+```py
+def bool_instance() -> bool:
+    return True
+
+flag = bool_instance()
+
+class Foo:
+    def __iadd__(self, other: int) -> str:
+        return "Hello, world!"
+
+if flag:
+    f = Foo()
+else:
+    f = 42.0
+f += 12
+
+# TODO(charlie): This should be `str | int`.
```

Comment:
Shouldn't this be `str | float`?

Fixing this doesn't have a dependency on removing `Type::Unbound`, right?

Reply:
Yeah, I think this is a TODO in binary operation handling, but... I included it anyway.